src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj

```diff
@@ -1,64 +1,117 @@
+    (.start scheduler)
+    scheduler))
 
-(defn wrap-with-error-logging
-  "Returns a function that will invoke 'f' inside a try/catch block.  If an
-  error occurs during execution of 'f', it will be logged and re-thrown."
-  [f]
-  (fn []
-    (try
-      (f)
-      (catch Throwable t
-        (log/error t (i18n/trs "scheduled job threw error"))
-        (throw t)))))
+(defn build-executable-job
+  ([f job-name group-name] (build-executable-job f job-name group-name {}))
+  ([f job-name group-name options]
+   (let [jdm (JobDataMap.)
+         options (assoc options :job f)]
+     (.put jdm "jobData" options)
+     (-> (JobBuilder/newJob puppetlabs.trapperkeeper.services.scheduler.job)
+         (.withIdentity job-name group-name)
+         (.usingJobData jdm)
+         (.build)))))
 
 (defn interspaced
-  ; See docs on the service protocol,
-  ; puppetlabs.enterprise.services.protocols.scheduler
-  [n f pool]
-  (let [job (wrap-with-error-logging f)]
-    (-> (at-at/interspaced n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  [n f ^Scheduler scheduler group-name]
+  (try
+    (let [job-name (Key/createUniqueName group-name)
+           job (-> (build-executable-job f job-name group-name {:interval n}))
+           schedule (-> (SimpleScheduleBuilder/simpleSchedule))
+           trigger (-> (TriggerBuilder/newTrigger)
+                       (.withSchedule schedule)
+                       (.startNow)
+                       (.build))]
+      (.scheduleJob scheduler job trigger)
+      (.getJobKey trigger))
+    (catch SchedulerException e
+      ; this can occur if the interface is being used while the scheduler is shutdown
+      (log/error e (i18n/trs "Failed to schedule job")))))
 
 (defn after
-  ; See docs on the service protocol,
-  ; puppetlabs.enterprise.services.protocols.scheduler
-  [n f pool]
-  (let [job (wrap-with-error-logging f)]
-    (-> (at-at/after n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  [n f ^Scheduler scheduler group-name]
+  (try
+    (let [job-name (Key/createUniqueName group-name)
+          job (build-executable-job f job-name group-name)
+          future-date (Date. ^Long (+ (System/currentTimeMillis) n))
+          trigger (-> (TriggerBuilder/newTrigger)
+                      (.startAt future-date)
+                      (.build))]
+      (.scheduleJob scheduler job trigger)
+      (.getJobKey trigger))
+    (catch SchedulerException e
+      ; this can occur if the interface is being used while the scheduler is shutdown
+      (log/error e (i18n/trs "Failed to schedule job")))))
 
 (defn stop-job
-  "Gracefully stops the job specified by 'id'."
-  [id pool]
-  (at-at/stop id pool))
+  "Returns true, if the job was deleted, and false if the job wasn't found."
+  [^JobKey id ^Scheduler scheduler]
+  (try
+    (.deleteJob scheduler id)
```

Comment:
Should we interrupt this job as well?

Using stop-job in stop-all-jobs' do-seq would be nice, unless there's a good reason to keep them separate, in which case a comment would be great.

Reply:
The documented behavior of the service's `stop-job` is that it doesn't interrupt a job that is running.  I could augment the function with a flag to indicate if it should interrupt or not and reuse the function that way.  Thoughts?

Reply:
Oh yeah, that makes sense.

I don't think I'll remember that in another year when we come back to this next, but now that you say it I see how that was the meaning of the docs.

I don't think it's necessarily worth changing the arity for different consumers.

Reply:
We are planning on making some new additions to the interface later, and can look at exposing a flag to force interrupt at that point.  That would allow us to reuse the function more easily.